---
title: "Bulk RNA-Sequencing REMI Tutorial"
author: "Alice Yu, Christine Yeh"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Bulk RNA-Sequencing REMI Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Getting started with REMI
```{r}
# install.packages("devtools")
# library(devtools)
# install_github("plevritis-lab/remi")
library(remi); library(knitr);library(ggplot2);library(ggalluvial)
```

# Read in data
Input file is a log-transformed bulk flow-sorted RNA-sequencing dataset from Gentles et al. _Genome Biology_ (2020). Column names are labeled as samplename_celltype.
```{r}
data("u01adeno")
dat <- u01adeno
```

# Running REMI
There is a built in filtering step for REMI where users can choose to threshold their genes based on quantiles. The number 4 represents the number of bins the data is separated into. Genes in the lowest expression quantile will be removed. Users can also turn off the filtering option (filter = F), but setupData needs to run in order to make a REMI object for the algorithm. Note: the run time of REMI is proportional to the number of cell types of interest. 
```{r}
# list of cell types user wants to analyze 
celltypes <- c("I", "E", "F", "M")
# Create REMI object
exp.list <- setupData(dat, celltypes, 4, filter=T)
# RUNNING REMI
remi.obj <- remi(exp.list)
```

# REMI Output
REMI interactome table
```{r}
kable(head(remi.obj$interactome, 5), digits = 5)
```

Plotting option #1
```{r fig.width=8,fig.height=6, warning=FALSE}
# Defining colors for each cell type
grid.col <- c("F" = '#BB6A23', 
              "I" = "#195188", 
              "M" = "#00503C", 
              "E" = "#AC371F")
# Plotting an alluvial plot showing proportion of cell types interacting
REMIPlot(remi.obj, 
         type="alluvial", 
         grid.col=grid.col)
```

Plotting option #2
```{r}
# Plotting a chord diagram showing proportion of cell types interacting
REMIPlot(remi.obj, 
         type="chord", 
         grid.col=grid.col)
```

# Calculate significance of an edge
Enter in ligand and receptor of choice and which community number the pair is located in. The p-value is calculated with respect to the community. Please add cell type of ligand into ligand name (celltype_ligand) and cell type of receptor into receptor name (celltype_receptor) as input. Note: this step may take a while and we recommend only running it for LR pairs of interest.
```{r}
#pval <- calculateSignificance(remi.obj,
                              # l.chosen="F_CTGF",
                              # r.chosen="F_LRP6",
                              # comm.chosen="50")
```



